node/network/bridge/src/lib.rs

```diff
@@ -141,355 +138,566 @@ struct PeerData {
+							} else {
+								tracing::trace!(
+									target: LOG_TARGET,
+									action = "PeerMessages",
+									peer = ?remote,
+									num_validation_messages = %v_messages.len(),
+									num_collation_messages = %c_messages.len()
+								);
+
+								if !v_messages.is_empty() {
+									let (events, reports) = handle_peer_messages(
+										remote.clone(),
+										&mut shared.0.lock().validation_peers,
+										v_messages,
+									);
+
+									for report in reports {
+										network_service.report_peer(remote.clone(), report).await?;
+									}
+
+									dispatch_validation_events_to_all(events, &mut sender).await;
+								}
+
+								if !c_messages.is_empty() {
+									let (events, reports) = handle_peer_messages(
+										remote.clone(),
+										&mut shared.0.lock().collation_peers,
+										c_messages,
+									);
+
+									for report in reports {
+										network_service.report_peer(remote.clone(), report).await?;
+									}
+
+
+									dispatch_collation_events_to_all(events, &mut sender).await;
+								}
+							}
+						}
 					}
 				}
 			},
-			Action::PeerMessages(peer, v_messages, c_messages) => {
-				tracing::trace!(
-					target: LOG_TARGET,
-					action = "PeerMessages",
-					peer = ?peer,
-					num_validation_messages = %v_messages.len(),
-					num_collation_messages = %c_messages.len()
-				);
-
-				if !v_messages.is_empty() {
-					let events = handle_peer_messages(
-						peer.clone(),
-						&mut validation_peers,
-						v_messages,
-						&mut bridge.network_service,
-					).await?;
-
-					dispatch_validation_events_to_all(events, &mut ctx).await;
+			req_res_event = request_multiplexer.next().fuse() => match req_res_event {
+				None => return Err(UnexpectedAbort::RequestStreamConcluded),
+				Some(Err(err)) => {
+					sender.send_message(NetworkBridgeMessage::ReportPeer(
+						err.peer,
+						MALFORMED_MESSAGE_COST,
+					).into()).await;
 				}
-
-				if !c_messages.is_empty() {
-					let events = handle_peer_messages(
-						peer.clone(),
-						&mut collation_peers,
-						c_messages,
-						&mut bridge.network_service,
-					).await?;
-
-					dispatch_collation_events_to_all(events, &mut ctx).await;
+				Some(Ok(msg)) => {
+					sender.send_message(msg).await;
 				}
 			},
-			Action::SendMessage(msg) => ctx.send_message(msg).await,
 		}
+	}
+}
 
-		tracing::trace!(
-			target: LOG_TARGET,
-			elapsed = ?before_action_process.elapsed(),
-			"Processed action",
-		);
+/// Main driver, processing network events and messages from other subsystems.
+///
+/// THIS IS A HACK. We need to ensure we never hold the mutex across a `.await` boundary
+/// and `parking_lot` currently does not provide `Send`, which helps us enforce that.
+/// If this breaks, we need to find another way to protect ourselves.
+///
+/// ```compile_fail
+/// #use parking_lot::MutexGuard;
+/// #fn is_send<T: Send>();
+/// #is_send::<parking_lot::MutexGuard<'static, ()>();
+/// ```
```

Comment:
haven't look at the actual issue, but if we need an async-aware Mutex, there is https://docs.rs/futures/0.3.13/futures/lock/struct.Mutex.html?

Reply:
@ordian I think `parking_lot::Mutex` is fine here. To avoid deadlock the enforcement that we never use the mutex across a `.await` is helpful. It means we will never be waiting to unpark the thread for very long.